apps/public/public.go

```diff
@@ -558,10 +559,33 @@ func WithRedirectURI(redirectURI string) interface {
 	}
 }
 
+// WithOpenURL allows you to provide a function to open the browser to complete the interactive login, instead of launching the system default browser.
+func WithOpenURL(openURL func(url string) error) interface {
```

Comment:
Agreed with the functionality of the API, but maybe not with the form the API, i.e. can we have some form of property bag for all extensibility related to browser? Maybe something like 

`WithSystemBrowserOptions(SystemBrowserOptions)`

This will avoid having to add to the top-level API everytime an advanced scenario needs to be open.

Reply:
By the way in MSAL .NET we allow folks to also customize the success message, failure message etc. It's just a matter of time before these are asked here as well.

Reply:
I also prefer property bags for methods but we're committed to the individual function pattern. An option function taking a property bag would be unconventional, and awkward next to the options we already have--wouldn't their settings belong in the bag as well?

> By the way in MSAL .NET we allow folks to also customize the success message, failure message etc. It's just a matter of time before these are asked here as well.

See #357. If the contributor doesn't finish it, I will when I have the cycles.

Reply:
Also agreed with the functionality in principle. Does this design imply that the caller (i.e. app developer) is responsible for somehow detecting that the current environment has no browser, therefore needs to *proactively* provide a callback which (quoted from #314) "just print out the URL to the console and move on to polling".

For what it's worth, [in MSAL Python](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.22.0/msal/oauth2cli/authcode.py#L255-L273), we can attempt browser blindly, detects the error afterwards, and then (by default) show a hint on the console to end user, all without the app developer needing to know all those details.

Reply:
Another scenario is to detect if Edge is installed on a machine and call Edge. Edge has a lot of functionality built in, that ensures Device CAP is satisfied and others. It's the only way to satisfy device CAP on linux today, until MSALs will be able to call the broker.

Reply:
Discussed with @chlowell offline. We clarified that, the new callback mechanism in this PR does not address any specific scenario in itself. We provide a flexibility to allow app developers to DIY. The onus is on app developer to detect and handle browser on different scenarios.